src/runtime-rs/crates/shimv2/src/args.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{os::unix::fs::FileTypeExt, path::PathBuf};
+
+use anyhow::{anyhow, Context, Result};
+
+use crate::Error;
+
+/// Received command-line arguments or environment arguments
+/// from a shimv2 container manager such as containerd.
+///
+/// For defail information, please refer to the
+/// [shim v2 spec](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md).
+#[derive(Debug, Default, Clone)]
+pub struct Args {
+    /// the id of the container
+    pub id: String,
+    /// the namespace for the container
+    pub namespace: String,
+    /// the address of the containerd's main socket
+    pub address: String,
+    /// the binary path to publish events back to containerd
+    pub publish_binary: String,
+    ///  Abstract socket path to serve.
+    pub socket: String,
+    /// the path to the bundle to delete
+    pub bundle: String,
+    /// Whether or not to enable debug
+    pub debug: bool,
+}
+
+impl Args {
+    /// Check the shim argument object is vaild or not.
+    ///
+    /// The id, namespace, address and publish_binary are mandatory for START, RUN and DELETE.
+    /// And bundle is mandatory for DELETE.
+    pub fn validate(&mut self, should_check_bundle: bool) -> Result<()> {
+        if self.id.is_empty()
+            || self.namespace.is_empty()
+            || self.address.is_empty()
+            || self.publish_binary.is_empty()
+        {
+            return Err(anyhow!(Error::ArgumentIsEmpty(format!(
+                "id: {} namespace: {} address: {} publish_binary: {}",
+                &self.id, &self.namespace, &self.address, &self.publish_binary
+            ))));
+        }
+        if should_check_bundle && self.bundle.is_empty() {
+            return Err(anyhow!(Error::ArgumentIsEmpty("bundle".to_string())));
+        }
+
+        if Self::is_component_dangerous(&self.id) || Self::is_component_dangerous(&self.namespace) {
+            return Err(anyhow!(Error::DangenerousComponent(format!(
+                "id: {} namespace {}",
+                &self.id, &self.namespace
+            ))));
+        }
+
+        // Ensure `address` is a valid path.
+        let path = PathBuf::from(self.address.clone())
+            .canonicalize()
+            .context(Error::InvalidPath(self.address.clone()))?;
+        let md = path
+            .metadata()
+            .context(Error::FileGetMetadata(format!("{:?}", path)))?;
+        if !md.file_type().is_socket() {
+            return Err(Error::InvalidArgument).context("address is not socket");
+        }
+        self.address = path
+            .to_str()
+            .map(|v| v.to_owned())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Ensure `bundle` is a valid path.
+        if !self.bundle.is_empty() {
```

Comment:
when would bundle be empty if `should_check_bundle` is not satisfied? 
should we just check for this and return earlier along with the other empty checks where we return `ArgumentIsEmpty`?

Reply:
when `should_check_bundle` the bundle must not empty, has check before `if should_check_bundle && self.bundle.is_empty()`